File: src/data/courses/dsa/subarrays-subsets-recursion/backtracking/letter-combinations-phone-number/code/java/LetterCombinationsPhoneNumber.java
// Time: O(4^N × N), Space: O(N)

import java.util.*;

class Solution {
    private static final String[] MAPPING = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if (digits.isEmpty()) return result;
        backtrack(digits, 0, new StringBuilder(), result);
        return result;
    }
    
    private void backtrack(String digits, int index, StringBuilder current, List<String> result) {
        if (index == digits.length()) {
            result.add(current.toString());
            return;
        }
        
        String letters = MAPPING[digits.charAt(index) - '0'];
        for (char letter : letters.toCharArray()) {
            current.append(letter);
            backtrack(digits, index + 1, current, result);
            current.deleteCharAt(current.length() - 1);
        }
    }
}

public class LetterCombinationsPhoneNumber {
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.letterCombinations("23"));
    }
}

File: src/data/courses/dsa/subarrays-subsets-recursion/backtracking/letter-combinations-phone-number/code/python/letter-combinations-phone-number.py
# Time: O(4^N × N), Space: O(N)

def letter_combinations(digits):
    if not digits:
        return []
    
    mapping = {"2": "abc", "3": "def", "4": "ghi", "5": "jkl", 
               "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"}
    result = []
    
    def backtrack(index, current):
        if index == len(digits):
            result.append(current)
            return
        
        for letter in mapping[digits[index]]:
            backtrack(index + 1, current + letter)
    
    backtrack(0, "")
    return result

digits = "23"
print(letter_combinations(digits))
